### PR TITLE
Adding support for gitlab ci and none ci provider

### PR DESCRIPTION
**Pull request by costrouc**
Closes https://github.com/Quansight/qhub-cloud/issues/210
Closes https://github.com/Quansight/qhub-cloud/issues/406

### CODE REVIEW

**Comment by viniciusdc** [comment]
LGTM, @costrouc have you tested the `--ci-provider` subcommand when deploying Qhub? Also, what dou you think using Mamba to solve the env for Jupyterlab docker image?

File: qhub/template/{{ cookiecutter.repo_directory }}/.gitlab-ci.yml [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+  variables:
+    IMAGE_TAG: {{ '$CI_COMMIT_SHA' }}
+    IMAGE_NAME: {{ cookiecutter.project_name }}-{{ '$DOCKERFILE' }}
+  script:
+    - docker build -f "image/Dockerfile.{{ '$DOCKERFILE' }}" -t "$IMAGE_NAME:$IMAGE_TAG" image
+  only:
+    changes:
+      - "image/**"
+
+terraform-jupyterhub:
```
Comment (viniciusdc) [inline qhub/template/{{ cookiecutter.repo_directory }}/.gitlab-ci.yml:40]: @costrouc do you have any ideas on how I can test this on Gitlab (how to overcome the secrets here)?

File: qhub/template/{{ cookiecutter.repo_directory }}/.gitlab-ci.yml [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        - "infrastructure/**" 
+        - "environments/**"
+    
+    - if: $PREFECT
+      variables:
+        VARS: '-var="prefect_token=${{ secrets.PREFECT_TOKEN }}"'
+
+    - if: $PROVIDER == 'aws'
+      variables:
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
Comment (viniciusdc) [inline qhub/template/{{ cookiecutter.repo_directory }}/.gitlab-ci.yml:61]: I will change this `secrets.GITHUB_TOKEN` later, we need to match this with Gitlab keys right?

**Comment by viniciusdc** [comment]
Hi @costrouc when testing the terraform jobs I got some errors, similar to this one:
```python
Error: Invalid expression
  on infrastructure/aws.tf line 76, in module "kubernetes":
  76: {% for nodegroup, nodegroup_config in cookiecutter.amazon_web_services.node_groups.items() %}
Expected the start of an expression, but found an invalid expression token.

```

**Comment by viniciusdc** [comment]
> Hi @costrouc when testing the terraform jobs I got some errors, similar to this one:
> 
> ```python
> Error: Invalid expression
>   on infrastructure/aws.tf line 76, in module "kubernetes":
>   76: {% for nodegroup, nodegroup_config in cookiecutter.amazon_web_services.node_groups.items() %}
> Expected the start of an expression, but found an invalid expression token.
> ```

Maybe we might not need to run that... I don't know. There is also `image.yaml` to mimic

**Comment by viniciusdc** [comment]
@costrouc when you have some time, could you review the Jinja structures here? (pretty sure I messed this variables)

**Comment by viniciusdc** [comment]
I will update the terraform job asap, for the tests I was able to deploy with aws and gcp. I still need to test do and azure.

**Comment by costrouc** [comment]
Closing in favor of #505

**Pull request closed without merging**